src/staticModelAdapt.cpp

```diff
@@ -0,0 +1,128 @@
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// RcppSMC is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with RcppSMC.  If not, see <http://www.gnu.org/licenses/>.
+
+#include "smctc.h"
+
+//! \file
+//! \brief This file contains the function definitions for adaptation
+//! in SMC for static Bayesian models.
+
+namespace smc {
+    /// Computes the difference between the conditional ESS given the specified temperature difference and the desired conditional ESS.
+    double staticModelAdapt::CESSdiff(const arma::vec & logweight, const arma::vec & loglike, double tempDiff, double desiredCESS){
+        double logsum1 = stableLogSumWeights(logweight + tempDiff*loglike);
+        double logsum2 = stableLogSumWeights(logweight + 2*tempDiff*loglike);
+
+        return exp(log(static_cast<double>(logweight.n_rows)) + 2*logsum1 - logsum2) - desiredCESS;
+    }
+    
+    ///Performs the bisection method to find the temperature within (temp_curr,1) which gives the desired conditional ESS.
+    double staticModelAdapt::bisection(double curr, const arma::vec & logweight, const arma::vec & loglike, const double & desiredCESS){
+        double a = curr;
+        double b = 1;
+        double f_a = CESSdiff(logweight,loglike,a-curr,desiredCESS);
+        double f_b = CESSdiff(logweight,loglike,b-curr,desiredCESS);
+        if (f_a*f_b>0){
+            Rcpp::stop("Bisection method to choose the next temperature failed");
+        } else{
+            double m, f_m, err;
+            m = (a+b)/2.0;
+            f_m = CESSdiff(logweight,loglike,m-curr,desiredCESS);
+            err = 10;
+            while (err > 0.01){
+                if (f_m<0){
+                    b = m;
+                    f_b = f_m;
+                } else{
+                    a = m;
+                    f_a = f_m;
+                }
+                m = (a+b)/2.0;
+                f_m = CESSdiff(logweight,loglike,m-curr,desiredCESS);
+                err = abs(f_m);
+            }
+            return m;
+        }
+    }
+
+    /// Chooses the next temperature such that a desired conditional ESS is maintained.
+    ///
+    /// \param logweight An armadillo vector containing the logarithm of the current particle weights.
+    /// \param loglike An armadillo vector containing the log likelihood of the current particle values.
+    /// \param desiredCESS The target conditional ESS for the next temperature (generally fixed).
+    void staticModelAdapt::ChooseTemp(const arma::vec & logweight, const arma::vec & loglike, double desiredCESS) {
+        double temp_curr = temp.back();
+        if (CESSdiff(logweight,loglike,1-temp_curr,desiredCESS)>=0){
```

Comment:
Might it make sense for the comparison to be with -0.01 (or the tolerance) to be consistent with the inner workings of the bisection function?